flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -1538,6 +1538,41 @@ def fir_BoxTypeDescOp : fir_SimpleOneResultOp<"box_tdesc", [NoSideEffect]> {
 
 // Record and array type operations
 
+def fir_ArrayCoorOp : fir_Op<"array_coor", [NoSideEffect]> {
+  let summary = "Find the coordinate of an element of an array";
+
+  let description = [{
+    Compute the location of an element in an array when the shape of the
+    array is only known at runtime.
+
+    This operation is intended to capture all the runtime values needed to
+    compute the address of an array reference in a single high-level op. Given
+    the following Fortran input:
+    
+    ```fortran
+      real :: a(n,m)
+      ...
+      ... a(i,j) ...
+    ```
+    
+    One can use `fir.array_coor` to determine the address of `a(i,j)`.
+    
+    ```mlir
+      %dim = fir.gendims %one, %n, %one, %one, %m, %one : (...) -> !fir.dims<2>
+      %1 = fir.array_coor %a(%dim), %i, %j : (!fir.ref<!fir.array<?x?xf32>>, ...) -> !fir.ref<f32>
+    ```
+  }];
+
+  let arguments = (ins AnyReferenceLike:$ref, fir_DimsType:$dims,
+                   Variadic<AnyIntegerType>:$coor);
```

Comment:
or `AnyCoordinateType`?

Reply:
Well, coordinate allows the use of `!fir.field` for example. I think we want to restrict this op to only be used on array references.

Reply:
(`!fir.field` is the type of a field specifier to compute the offset of a field in a record at runtime.)